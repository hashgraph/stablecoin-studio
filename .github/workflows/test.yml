name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test