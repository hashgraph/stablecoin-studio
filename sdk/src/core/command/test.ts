/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
