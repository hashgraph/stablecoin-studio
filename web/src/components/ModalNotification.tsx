import {
	Button,
	Image,
	Link,
	Modal,
	ModalBody,
	ModalCloseButton,
	ModalContent,
	ModalFooter,
	ModalHeader,
	ModalOverlay,
	Spinner,
	Text,
} from '@chakra-ui/react';
import type { ModalProps } from '@chakra-ui/react';
import { useTranslation } from 'react-i18next';
import SUCCESS_ICON from '../assets/svg/success.svg';
import ERROR_ICON from '../assets/svg/error.svg';
import WARNING_ICON from '../assets/svg/warning.svg';

const SUCCESS = 'success';
const WARNING = 'warning';
const ERROR = 'error';

interface ModalNotificationProps extends Omit<ModalProps, 'children'> {
	description?: string | null;
	icon?: string;
	isOpen: boolean;
	onClose: () => void;
	title: string|null;
	variant?: 'error' | 'success' | 'warning' | 'loading';
	onClick?: () => void;
	closeButton?: boolean;
	errorTransactionUrl?: string | null;
}

const ModalNotification = (props: ModalNotificationProps) => {
	const {
		description,
		icon,
		isOpen,
		onClick,
		onClose,
		title,
		variant,
		closeButton = true,
		errorTransactionUrl,
		...othersProps
	} = props;
	const { t } = useTranslation('global');
	const isLoading = (variant && variant === 'loading') ?? false;

	const getIcon = () => {
		if (icon) return icon;
		switch (variant) {
			case SUCCESS:
				return SUCCESS_ICON;
			case WARNING:
				return WARNING_ICON;
			case ERROR:
				return ERROR_ICON;
		}
	};

	return (
		<Modal
			data-testid='modal-notification'
			isOpen={isOpen}
			onClose={onClose}
			isCentered
			blockScrollOnMount={false}
			{...othersProps}
		>
			<ModalOverlay />
			<ModalContent data-testid='modal-notification-content' pt='50' pb='50'>
				{closeButton && !isLoading && <ModalCloseButton />}
				{(icon || variant) && !isLoading && (
					<ModalHeader alignSelf='center' p='0'>
						<Image
							data-testid='modal-notification-icon'
							src={getIcon()}
							width='54px'
							height='54px'
						/>
					</ModalHeader>
				)}
				{isLoading && (
					<Spinner w='54px' h='54px' justifyContent='center' alignSelf={'center'} color='#C6AEFA' thickness='4px'/>
				)}
				<ModalBody textAlign='center' pt='14px'>
					<Text
						data-testid='modal-notification-title'
						fontSize='14px'
						fontWeight={700}
						lineHeight='16px'
						color='brand.black'
					>
						{title}
					</Text>
					{description && (
						<Text
							data-testid='modal-notification-description'
							mt='8px'
							fontSize='12px'
							fontWeight={400}
							lineHeight='16px'
							color='brand.gray'
						>
							{description}
						</Text>
					)}
					{errorTransactionUrl && (
						<Link href={errorTransactionUrl} isExternal textDecoration='underline'>
							{t('common.see-transaction')}
						</Link>
					)}
				</ModalBody>
				{!isLoading && (
					<ModalFooter alignSelf='center' pt='24px' pb='0'>
						<Button
							data-testid='modal-notification-button'
							onClick={onClick || onClose}
							variant='primary'
						>
							{t('common.accept')}
						</Button>
					</ModalFooter>
				)}
			</ModalContent>
		</Modal>
	);
};

export default ModalNotification;
