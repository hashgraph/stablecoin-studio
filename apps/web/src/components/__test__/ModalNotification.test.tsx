import userEvent from '@testing-library/user-event';
import ModalNotification from '../ModalNotification';
import { render } from '../../test/index';
import HEDERA_LOGO from '../../assets/png/hedera-hbar-logo.png';

const defaultProps = {
	title: 'Title',
	isOpen: true,
	onClose: jest.fn(),
};

describe(`<${ModalNotification.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<ModalNotification {...defaultProps} />);

		expect(component.asFragment()).toMatchSnapshot('default');
	});

	test('should not has icon without icon or variant props', () => {
		const component = render(<ModalNotification {...defaultProps} />);

		const icon = component.queryByTestId('modal-notification-icon');
		expect(icon).not.toBeInTheDocument();
	});

	test('should render icon passed by prop', () => {
		const component = render(<ModalNotification icon={HEDERA_LOGO} {...defaultProps} />);

		const icon = component.getByTestId('modal-notification-icon');
		expect(icon).toBeInTheDocument();
	});

	test('should render success icon if variant is success', () => {
		const component = render(<ModalNotification variant='success' {...defaultProps} />);

		const successIcon = component.getByTestId('modal-notification-icon');
		expect(successIcon).toBeInTheDocument();
		expect(successIcon).toHaveAttribute('src', 'success.svg');

		expect(component.asFragment()).toMatchSnapshot('success');
	});

	test('should render error icon if variant is error', () => {
		const component = render(<ModalNotification variant='error' {...defaultProps} />);

		const successIcon = component.getByTestId('modal-notification-icon');
		expect(successIcon).toBeInTheDocument();
		expect(successIcon).toHaveAttribute('src', 'error.svg');

		expect(component.asFragment()).toMatchSnapshot('error');
	});

	test('should render title', () => {
		const component = render(<ModalNotification {...defaultProps} />);

		const title = component.getByTestId('modal-notification-title');
		expect(title).toBeInTheDocument();
		expect(title).toHaveTextContent(defaultProps.title);
	});

	test('should not render description if it is not on props', () => {
		const component = render(<ModalNotification {...defaultProps} />);

		const description = component.queryByTestId('modal-notification-description');
		expect(description).not.toBeInTheDocument();
	});

	test('should render description if it is on props', () => {
		const descriptionProp = 'notification description';
		const component = render(<ModalNotification description={descriptionProp} {...defaultProps} />);

		const description = component.getByTestId('modal-notification-description');
		expect(description).toBeInTheDocument();
		expect(description).toHaveTextContent(descriptionProp);

		expect(component.asFragment()).toMatchSnapshot('with-description');
	});

	test('should has an Accept button', () => {
		const component = render(<ModalNotification {...defaultProps} />);

		const button = component.getByTestId('modal-notification-button');
		expect(button).toBeInTheDocument();
	});

	test('Accept button should close modal by default', async () => {
		const component = render(<ModalNotification {...defaultProps} />);

		const button = component.getByTestId('modal-notification-button');
		await userEvent.click(button);

		expect(defaultProps.onClose).toHaveBeenCalled();
	});

	test('Accept button should call function passed by onClick prop', async () => {
		const acceptButton = jest.fn();
		const component = render(<ModalNotification onClick={acceptButton} {...defaultProps} />);

		const button = component.getByTestId('modal-notification-button');
		await userEvent.click(button);

		expect(acceptButton).toHaveBeenCalled();
	});
});
