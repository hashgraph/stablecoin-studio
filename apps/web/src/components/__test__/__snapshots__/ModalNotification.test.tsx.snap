// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalNotification /> should render correctly: default 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<ModalNotification /> should render description if it is on props: with-description 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<ModalNotification /> should render error icon if variant is error: error 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<ModalNotification /> should render success icon if variant is success: success 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
